airflow/models/dag.py

```diff
@@ -522,7 +522,7 @@ def __init__(
         self.has_on_success_callback = self.on_success_callback is not None
         self.has_on_failure_callback = self.on_failure_callback is not None
 
-        self.doc_md = doc_md
+        self.doc_md = pathlib.Path(doc_md).read_text() if pathlib.Path(str(doc_md or '')).is_file() else doc_md
```

Comment:
Instead of doing heuristics like this, I wonder if it’s easier to simply _require_ the user to pass in a `Path` object, and do this here:

```python
if isinstance(doc_md, pathlib.Path):
    self.doc_md = doc_md.read_text("utf-8")
else:
    self.doc_md = doc_md
```

Reply:
I was not able to find the logic used for .sql files in the SQL operators like the [here](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/operators/sql.py#L802).

While testing I was not able to just pass "my_dag.py" but I had to pass a Path object, like you were mentioning, which is in my opinion more cumbersome and inconsistent (with what we do for .sql files)

Reply:
Uhm... it looks like it is a jinja2 feature and trying to get_template_env with something like this:
```python
    def get_doc_md(self, doc_md: str) -> str:
        if doc_md is None or not doc_md.endswith('.md'):
            return doc_md

        env = self.get_template_env(force_sandboxed=False)
        
        return env.loader.get_source(env, doc_md)
```
 gives me a weird error
 ```
 Broken DAG: [/files/dags/dag.py] Traceback (most recent call last):
  File "/opt/airflow/airflow/serialization/serialized_objects.py", line 286, in validate_schema
    cls._json_schema.validate(serialized_obj)
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 353, in validate
    raise error
jsonschema.exceptions.ValidationError: 
```
Followed by the content of my `.md` file

Reply:
found the problem :)

Reply:
it now works with a dag defined as:
```python
my_dag = DAG(dag_id="sample_dag",
                     start_date=datetime(2020, 4, 29),
                     schedule_interval="0 0 * * *",
                     doc_md='dag.md'
             )
```

Because it is handled via Jinja, it should be able to handle templating too :)

Reply:
I figured out how to render the `.md` file as a Jinja template.

Unfortunately the `doc_md` template at DAG level cannot access Airflow macros because there is no  context (like for the templated fields for tasks). As mentioned [here](https://github.com/apache/airflow/discussions/25396#discussioncomment-3280373).

Of course a macro like `{{ ds }}` has no meaning at DAG level, but accessing Airflow Variables or Connection could be useful.

If anyone has an idea on how to add the context, I can try to do it :)

Reply:
we can't have context because the moment we read it, the context does not exist yet.. We could potentially add a specially crafted context with only a few things that we do have access (like dag).

Reply:
Yes, I was thinking about something like that, but I have no idea how big would be the effort and the ROI of it :)

I would say, let's see if anyone else in the community is interested in such kind of feature

Reply:
Yes.